bindings/go/oci/credentials/docker_config.go

```diff
@@ -0,0 +1,179 @@
+	registeredPort, portInIdentity := identity[runtime.IdentityAttributePort]
+
+	return func(ctx context.Context, hostport string) (auth.Credential, error) {
+		actualHost, actualPort, err := net.SplitHostPort(hostport)
+		if err != nil {
+			return auth.Credential{}, fmt.Errorf("failed to split host and port: %w", err)
+		}
+		hostMismatch := hostInIdentity && registeredHostname != actualHost
+		portMismatch := portInIdentity && registeredPort != actualPort
+		if hostMismatch || portMismatch {
+			return auth.EmptyCredential, nil
+		}
+		return credential, nil
+	}
+}
+
+// ResolveV1DockerConfigCredentials resolves credentials from a Docker configuration
+// for a given identity. It supports both file-based and in-memory Docker configurations.
+//
+// The function will:
+//   - Load credentials from the specified Docker config source
+//   - Match the credentials against the provided identity
+//   - Return a map of credential key-value pairs
+func ResolveV1DockerConfigCredentials(ctx context.Context, dockerConfig credentialsv1.DockerConfig, identity runtime.Identity) (map[string]string, error) {
+	credStore, err := getStore(ctx, dockerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve credentials store: %w", err)
+	}
+
+	hostname := identity[runtime.IdentityAttributeHostname]
+	if hostname == "" {
+		return nil, fmt.Errorf("missing %q in identity", runtime.IdentityAttributeHostname)
+	}
+
+	cred, err := credStore.Get(ctx, hostname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get credentials for %q: %w", hostname, err)
+	}
+	credentialMap := map[string]string{}
+	if v := cred.Username; v != "" {
+		credentialMap[CredentialKeyUsername] = v
+	}
+	if v := cred.Password; v != "" {
+		credentialMap[CredentialKeyPassword] = v
+	}
+	if v := cred.AccessToken; v != "" {
+		credentialMap[CredentialKeyAccessToken] = v
+	}
+	if v := cred.RefreshToken; v != "" {
+		credentialMap[CredentialKeyRefreshToken] = v
+	}
+
+	return credentialMap, nil
+}
+
+// getStore creates a credential store based on the provided Docker configuration.
+// It supports three modes of operation:
+//   - Default mode: Uses system default Docker config locations
+//   - Inline config: Uses a provided JSON configuration string
+//   - File-based: Uses a specified Docker config file
+//
+// The function handles:
+//   - Shell expansion for file paths (e.g., ~ for home directory)
+//   - Temporary file creation for inline configurations
+//   - Integration with the native host credential store
+func getStore(ctx context.Context, dockerConfig credentialsv1.DockerConfig) (remotecredentials.Store, error) {
+	var store remotecredentials.Store
+	var err error
+	switch {
+	case dockerConfig.DockerConfigFile == "" && dockerConfig.DockerConfig == "":
+		slog.InfoContext(ctx, "attempting to load docker config from default locations or native host store")
+		store, err = remotecredentials.NewStoreFromDocker(remotecredentials.StoreOptions{
```

Comment:
I don't particularly like overwriting error values. But it's fine unless you think that this code section here could possibly be called concurrently. If the answer is yes, that means that this is a potential exploit. ( the error value can be overwritten concurrently, and possibly could result in a nil error continuing a segment of code that you didn't wish for ).

Reply:
agreed. i split this method to make it more readable now please double check! :)

Reply:
thanks! <3